libyul/backends/wasm/WordSizeTransform.cpp

```diff
@@ -0,0 +1,275 @@
+				else if (assignment.value->type() == typeid(Identifier))
+				{
+					yulAssert(assignment.variableNames.size() == 1, "");
+					auto newRhs = expandIdentifierValue(boost::get<Identifier>(*assignment.value));
+					YulString lhsName = assignment.variableNames[0].name;
+					yulAssert(m_variableMapping.find(lhsName) != m_variableMapping.end(), "");
+					vector<Statement> ret;
+					for (int i = 0; i < 4; i++)
+						ret.push_back(
+							Assignment{
+								assignment.location,
+								{Identifier{assignment.location, m_variableMapping[lhsName][i]}},
+								std::move(newRhs[i])
+							}
+						);
+					return ret;
+				}
+				else if (assignment.value->type() == typeid(Literal))
+				{
+					yulAssert(assignment.variableNames.size() == 1, "");
+					auto newRhs = expandLiteralValue(boost::get<Literal>(*assignment.value));
+					YulString lhsName = assignment.variableNames[0].name;
+					yulAssert(m_variableMapping.find(lhsName) != m_variableMapping.end(), "");
+					vector<Statement> ret;
+					for (int i = 0; i < 4; i++)
+						ret.push_back(
+							Assignment{
+								assignment.location,
+								{Identifier{assignment.location, m_variableMapping[lhsName][i]}},
+								std::move(newRhs[i])
+							}
+						);
+					return ret;
+				}
+			}
+			else
+				visit(_s);
+			return boost::none;
+		}
+	);
+}
+
+array<YulString, 4> WordSizeTransform::generateNewNames(YulString const& _s)
+{
+	yulAssert(m_variableMapping.find(_s) == m_variableMapping.end(), "");
+	array<YulString, 4> ret;
+	for (int i = 0; i < 4; i++)
+	{
+		auto newName = m_nameDispenser.newName(_s);
+		m_variableMapping[_s][i] = newName;
+		ret[i] = newName;
+	}
+	return ret;
+}
+
+array<unique_ptr<Expression>, 4> WordSizeTransform::expandIdentifierValue(Identifier& _id)
+{
+	array<unique_ptr<Expression>, 4> ret;
+	auto iter = m_variableMapping.find(_id.name);
+	yulAssert(iter != m_variableMapping.end(), "");
+	for (int i = 0; i < 4; i++)
+		ret[i] = make_unique<Expression>(Identifier{_id.location, iter->second[i]});
+	return ret;
+}
+
+array<unique_ptr<Expression>, 4> WordSizeTransform::expandLiteralValue(Literal& _literal)
+{
+	yulAssert(_literal.kind == LiteralKind::Number, "Non-numeric literal unimplemented.");
+	u256 val = u256(_literal.value.str());
+	array<unique_ptr<Expression>, 4> ret;
+	for (int i = 3; i >= 0; i--)
+	{
+		u256 currentVal = val & u256("0x" + string(16, 'F')); // obtain least significant 64 bits
+		val >>= 64;
+		ret[i] = make_unique<Expression>(
+			Literal{
+				_literal.location,
+				LiteralKind::Number,
+				YulString(currentVal.str()),
+				"u64"_yulstring
+			}
+		);
+	}
+	return ret;
+}
+
+void WordSizeTransform::rewriteFunctionCallArguments(vector<Expression>& _args)
+{
+	iterateReplacing(
+		_args,
+		[&](Expression& _e) -> boost::optional<vector<Expression>>
+		{
+			vector<Expression> ret;
+			if (_e.type() == typeid(Identifier))
+			{
+				Identifier id = boost::get<Identifier>(_e);
+				for (auto& v: expandIdentifierValue(id))
```

Comment:
Same here - I think a function that just takes an `Expression` would be easier to use here.

Reply:
I think current version is probably more readable.

Reply:
I mean, we can probably only save several LoC by combining `expandLiteralValue()` and `expandIdentifierValue()`

Reply:
Yes, i'm proposing to combine `expandLiteralValue` and `expandIdentifierValue`. The first thing I do when I see this code is that I wonder "What is the difference between the first and the second block? Are Literals and Identifiers handled differently?". If the code is the same and the difference is only in the very "bottom", I won't have to check this.

Reply:
Ok :+1: , force pushed